edge-util/src/Microsoft.Azure.Devices.Edge.Storage.RocksDb/disk/DiskSpaceChecker.cs

```diff
@@ -0,0 +1,150 @@
+            this.inner = Preconditions.CheckNotNull(inner, nameof(inner));
+        }
+
+        public bool IsFull => this.inner.DiskStatus == DiskStatus.Full;
+
+        public static DiskSpaceChecker Create(string storageFolder, double thresholdPercentage, TimeSpan checkFrequency)
+        {
+            Option<DriveInfo> drive = GetMatchingDrive(storageFolder);
+            Option<string> driveName = drive.Map(d => d.Name);
+            DiskSpaceCheckerBase inner = driveName.Map(d => new ThresholdDiskSpaceChecker(d, thresholdPercentage, checkFrequency, Events.Log) as DiskSpaceCheckerBase)
+                .GetOrElse(() => new DummyDiskSpaceChecker(checkFrequency, Events.Log));
+            var diskSpaceChecker = new DiskSpaceChecker(storageFolder, driveName, checkFrequency, inner);
+            Events.Created(storageFolder);
+            return diskSpaceChecker;
+        }
+
+        public void SetThresholdPercentage(int thresholdPercentage)
+        {
+            string drive = this.drive.Expect(() => new InvalidOperationException("Cannot set disk usage threshold percentage since drive is not known"));
+            lock (this.updateLock)
+            {
+                Events.SetMaxPercentageUsage(thresholdPercentage, drive);
+                this.inner = new ThresholdDiskSpaceChecker(drive, thresholdPercentage, this.checkFrequency, Events.Log);
+            }
+        }
+
+        public void SetMaxDiskUsageSize(long maxDiskUsageBytes)
+        {
+            lock (this.updateLock)
+            {
+                Events.SetMaxSizeDiskSpaceUsage(maxDiskUsageBytes, this.storageFolder);
+                this.inner = new FixedSizeSpaceChecker(this.storageFolder, maxDiskUsageBytes, this.checkFrequency, Events.Log);
+            }
+        }
+
+        internal static Option<DriveInfo> GetMatchingDrive(string storageFolder)
+        {
+            DriveInfo match = null;
+            var drives = new DriveInfo[0];
+            try
+            {
+                drives = DriveInfo.GetDrives();
+                int segmentCount = -1;
+                foreach (DriveInfo drive in drives)
+                {
+                    if (drive.IsReady)
+                    {
+                        if (storageFolder.StartsWith(drive.Name, StringComparison.OrdinalIgnoreCase))
+                        {
+                            string[] segments = drive.Name.Split(new[] { Path.DirectorySeparatorChar }, StringSplitOptions.RemoveEmptyEntries);
+                            if (segments.Length > segmentCount)
+                            {
+                                segmentCount = segments.Length;
+                                match = drive;
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Events.ErrorGettingMatchingDrive(storageFolder, e);
+            }
+
+            if (match != null)
```

Comment:
I think this would be easier to read as `if (match == null) { Events.NoMatchingDriveFound(); return Option.None<>(); } Events.FoundDrive(); return Option.Maybe();` so that the two "success" branch statements are together.